### PR TITLE
BETA 4

### PR DESCRIPTION
**Pull request by hoffmabc**

### CODE REVIEW

**Comment by acrisci** [comment]
I would like to get involved with this project, but It's really hard to follow development when a lot of code is being tied up in a fork. Why is this not being incrementally merged with the develop branch?

What can I do to speed up the process of getting this mainlined? Can you give me some instructions on how to test what changes you have made?

**Comment by dionyziz** [comment]
@acrisci This is a major rewrite of our networking code. It moves from using TCP to using UDP so that network connectivity can be improved, and the ZeroMQ dependency is being dropped. It also introduces UDP hole punching and other important networking improvements.

I agree with you that it's best to merge incrementally into the develop branch. However, this particular PR is hard to merge incrementally, as it is a major deviation from what we've been doing in the past, and a redesign of our network layer. So it has to be a large PR.

I hope @hoffmabc will rebase this into commits that make sense and also possibly split up unrelated changes into a separate PR when time permits. As you see, this is marked as WIP work-in-progress, so it will get some love and cleanup as it progresses.

Thanks for your interest in contributing :)

File: installers/ubuntu/DEBIAN/control [hunk 1]
```diff
@@ -8,7 +8,7 @@ Build-Depends:
 Standards-Version: 3.9.5
 Homepage: https://openbazaar.org/
 Architecture: all
-Depends: python, python2.7, python-pip, python-dev, libjpeg-dev, sqlite3, libssl-dev, openssl, build-essential, libzmq3, python-zmq, tor, privoxy, rng-tools, lintian, libjs-jquery
+Depends: python, python2.7, python-pip, python-dev, libjpeg-dev, sqlite3, libssl-dev, openssl, build-essential, rng-tools, lintian, libjs-jquery
```
Comment (abderraouf-adjal) [inline installers/ubuntu/DEBIAN/control]: You can remove `python2.7` because `python` get python2 ?

File: util/docker/logelk/ob/Dockerfile [hunk 2]
```diff
@@ -5,7 +5,7 @@ MAINTAINER Y12STUDIO <y12studio@gmail.com>
 # package for OB
 #
 RUN apt-get -y update
-RUN apt-get install -y build-essential python-zmq rng-tools
+RUN apt-get install -y build-essential rng-tools
 RUN apt-get install -y python-dev g++ libjpeg-dev sqlite3 openssl
```
Comment (abderraouf-adjal) [inline util/docker/logelk/ob/Dockerfile]: Can `apt-get install` cmd be in one line?
And the package dependencies looks different from `DEBIAN/control` file, for eg. `build-essential` get `g++`.

File: installers/ubuntu/DEBIAN/control (hunk 1 above)
Comment (Renelvon) [inline installers/ubuntu/DEBIAN/control]: If `python` starts pointing to `python3`, there will be problems. So, it's best to keep `python2.7` and remove `python`.

File: util/docker/logelk/ob/Dockerfile (hunk 2 above)
Comment (Renelvon) [inline util/docker/logelk/ob/Dockerfile]: Neither if these files is up-to-date. Better check `configure.sh` for dependencies.

File: node/connection.py [hunk 3]
```diff
@@ -1,136 +1,257 @@
...
+        @self._rudp_connection._sender.ee.on('timeout')
+        def on_timeout(data):  # pylint: disable=unused-variable
+            self.log.debug('Node Sender Timed Out')
+            #self.transport.dht.remove_peer(self.guid)
+
+        @self._rudp_connection.ee.on('data')
+        def handle_recv(msg):  # pylint: disable=unused-variable
+
+            self.log.debug('Got the whole message: %s', msg.get('payload'))
+            payload = msg.get('payload')
```
Comment (gasull) [inline node/connection.py]: Why not swapping these two lines so you use ```payload``` in the line that is now at the top?  I.e.:
```Python
payload = msg.get('payload')
self.log.debug('Got the whole message: %s', payload)
```

File: node/connection.py [hunk 4]
```diff
@@ -1,136 +1,257 @@
...
+
+            if payload[:1] == '{':
+                try:
+                    payload = json.loads(msg.get('payload'))
```
Comment (gasull) [inline node/connection.py]:
```Python
payload = json.loads(payload)
```

File: node/connection.py [hunk 5]
```diff
@@ -1,136 +1,257 @@
...
+                    self.transport.listener._on_raw_message(payload)
+                    return
+                except Exception as e:
+                    self.log.debug('Problem with serializing: %s', e)
+            else:
+                try:
+                    payload = msg.get('payload').decode('hex')
```
Comment (gasull) [inline node/connection.py]:
```Python
payload = payload.decode('hex')
```

File: node/connection.py [hunk 6]
```diff
@@ -1,136 +1,257 @@
...
+                    self.transport.listener._on_raw_message(payload)
+                except Exception as e:
+                    self.log.debug('not yet %s', e)
+                    self.transport.listener._on_raw_message(msg.get('payload'))
+
+
+    def send_ping(self):
+        self.sock.sendto('ping', (self.hostname, self.port))
+        return True
```
Comment (gasull) [inline node/connection.py]: Why are we returning ```True```?

File: node/connection.py [hunk 7]
```diff
@@ -1,136 +1,257 @@
...
+                        self.log.debug('Found restricted NAT client')
+                        self.transport.start_mediation(self.guid)
+                    if self.nat_type == 'Full Cone' and not self.relaying:
+                        self.send_to_rudp(serialized)
+                        return
+                    else:
+                        self.log.debug('Relay through seed')
+                        # self.transport.relay_message(serialized, self.guid)
+                        self.send_to_rudp(serialized)
+                        return
```
Comment (gasull) [inline node/connection.py]: Modest proposal to reduce the number of lines and improve readibility a bit:
```Python
found_seed_relaying = (self.relaying or self.nat_type != 'Full Cone' or
                       (self.reachable and
                        self.transport.nat_type == 'Symmetric NAT'))
found_restricted_nat = (not self.relaying and not self.punching and
                        self.nat_type == 'Restric NAT')

if found_seed_relaying:
    # Relay through seed server
    self.log.debug('Relay through seed')
    # self.transport.relay_message(serialized, self.guid)
elif found_restricted_nat:
    self.log.debug('Found restricted NAT client')
    self.transport.start_mediation(self.guid)

self.send_to_rudp(serialized)
```
